control/statesp.py

```diff
@@ -1227,12 +1227,110 @@ def dcgain(self, warn_infinite=False):
         return self(0, warn_infinite=warn_infinite) if self.isctime() \
             else self(1, warn_infinite=warn_infinite)
 
+    def dynamics(self, *args):
+        """Compute the dynamics of the system
+
+        Given input `u` and state `x`, returns the dynamics of the state-space
+        system. If the system is continuous, returns the time derivative dx/dt
+
+            dx/dt = A x + B u
+
+        where A and B are the state-space matrices of the system. If the
+        system is discrete-time, returns the next value of `x`:
+
+            x[t+dt] = A x[t] + B u[t]
+
+        The inputs `x` and `u` must be of the correct length for the system.
+
+        The calling signature is ``out = sys.dynamics(t, x[, u])``
+        The first argument `t` is ignored because :class:`StateSpace` systems
+        are time-invariant. It is included so that the dynamics can be passed
+        to most numerical integrators, such as scipy's `integrate.solve_ivp` and
+        for consistency with :class:`IOSystem` systems.
+
+        Parameters
+        ----------
+        t : float (ignored)
+            time
+        x : array_like
+            current state
+        u : array_like (optional)
+            input, zero if omitted
+
+        Returns
+        -------
+        dx/dt or x[t+dt] : ndarray
+        """
+        if len(args) not in (2, 3):
+            raise ValueError("received" + len(args) + "args, expected 2 or 3")
+
+        x = np.reshape(args[1], (-1, 1)) # force to a column in case matrix
+        if np.size(x) != self.nstates:
+            raise ValueError("len(x) must be equal to number of states")
+
+        if len(args) == 2: # received t and x, ignore t
+            return self.A.dot(x).reshape((-1,)) # return as row vector
+        else: # received t, x, and u, ignore t
+            u = np.reshape(args[2], (-1, 1)) # force to a column in case matrix
+            if np.size(u) != self.ninputs:
+                raise ValueError("len(u) must be equal to number of inputs")
+            return self.A.dot(x).reshape((-1,)) \
+                 + self.B.dot(u).reshape((-1,)) # return as row vector
```

Comment:
Is there a reason to do it this way compared to defining the call signature as `dynamics(self, t, x, u=0)` and then having the function be something like
```
return self.A @ x if u == 0 else self.A @ x + self.B @ u
```
(perhaps with some dimension checking for user friendliness).

Reply:
good idea ... i think that would be much cleaner